docs/RFCS/certificate_rotation.md

```diff
@@ -0,0 +1,293 @@
+multiple files for node and client certificates due to changing keys.
+
+We thus need to be able to specify a storage location containing an arbitrary number of
+certificates and keys.
+
+We propose a single directory containing all relevant certificates and private keys for
+the current node or client.
+
+## Directory
+
+The directory can be specified by command-line flag `--certs-dir` or environment variable
+`COCKROACH_CERTS_DIR`, with a default value of `cockroach-certs`.
+
+The permissions on the directory cannot be more permissive than `drwx------`.
+Directory ownership will not be checked (too much incompatibility across filesystems and
+architectures), relying instead on permissions to list the directory contents.
+
+## Files
+
+File requirements are as follows:
+* all files must be in the `COCKROACH_CERTS_DIR` directory, sub-directories are ignored
+* all files must be plain files, no sym-links allowed
+* maximum permissions for certificates: `-rwxr-xr-x`
+* maximum permissions for keys: `-rwx------`
+* all files are PEM-encoded
+
+File types and functions are determined automatically into categories by
+checking the following PEM or certificate information:
+
+| `pem.Block.Type` | `IsCA` | `ExtendedKeyUsage` | `Subject.CommonName` | Category |
+|---|---|---|---|---|
+| `CERTIFICATE` | `TRUE` | `CertSign | ContentCommitment` | Any | CA Certificate |
+| `CERTIFICATE` | `FALSE` | `ServerAuth | ClientAuth` | `Node` | Node certificate |
+| `CERTIFICATE` | `FALSE` | `ClientAuth` | not `Node` | Client certificate |
+| `PRIVATE KEY` | | | | Private Key |
+
+Any other combination (eg: `ExtendedKeyUsage = ServerAuth | ClientAuth` and `User = foo`) is invalid.
+
+Certificates and private keys can be matched by checking public-key components:
+* `rsa.PublicKey.N.Cmp(rsa.PrivateKey.N) == 0` for RSA keys.
+* `ecdsa.PublicKey.X.Cmp(ecdsa.PrivateKey.X) == 0 && ecdsa.PublicKey.Y.Cmp(ecdsa.PrivateKey.Y) == 0` for ecdsa keys.
+
+Details of PEM parsing and key checking can be found in [tls.X509KeyPair](https://golang.org/src/crypto/tls/tls.go?s=5915:5986#L183)
+
+CA certificates can be added to the certificate pool with little extra validation.
+
+Node certificates must be checked for validity. Specifically:
+* we must have a valid certificate/private key pair
+* the certificate must currently be valid (`Not Before < time.Now() < Not After`)
+* the certificate must be signed by one of the CA certificates on this node
+
+The last condition is an attempt to avoid loading a certificate that may not be verifiable
+by other nodes or clients. If we do not have the right CA, chances are someone else does no either.
+
+We may need to set a timer for certificates that have not reached their `Not Before` date, otherwise
+we would need to trigger a second refresh.
+
+## Triggering a reload
+
+Running nodes can be told to re-read the certificate directory by issuing a `SIGHUP` to the process.
```

Comment:
I think here we should also say that the node will also spontaneously check for updated certificates at regular intervals, perhaps once a day, and certainly some time before the earliest expiration in the future.